lib/index.js

```diff
@@ -0,0 +1,333 @@
+        return callback(new BedrockError(
+          'Failed to find proof state due to internal error.',
+          'InternalError',
+          {proof: id, httpStatusCode: 500, 'public': true}, err));
+      }
+      if(!record) {
+        return callback(new BedrockError(
+          'Proof state not found.',
+          'NotFound',
+          {proof: id, httpStatusCode: 404, 'public': true}));
+      }
+      callback(null, record.state, record.meta);
+    });
+}
+
+function _insert(id, state, callback) {
+  var now = Date.now();
+  var record = {
+    id: database.hash(id),
+    meta: {
+      created: now,
+      updated: now
+    },
+    state: state
+  };
+  database.collections.jwtSecrets.insert(
+    record, database.writeOptions, function(err, result) {
+      if(err) {
+        if(database.isDuplicateError(err)) {
+          return callback(new BedrockError(
+            'The proof state is a duplicate and could not be added.',
+            'DuplicateProofState', {
+              proof: id,
+              httpStatusCode: 409,
+              'public': true
+            }));
+        }
+        logger.error('Mongo error when trying to insert proof state.', err);
+        return callback(new BedrockError(
+          'Failed to insert proof state due to internal error.',
+          'InternalError',
+          {proof: id, httpStatusCode: 500, 'public': true},
+          err));
+      }
+      callback(null, result.ops[0]);
+    });
+}
+
+function _update(id, state, options, callback) {
+  if(typeof options === 'function') {
+    callback = options;
+    options = {};
+  }
+  options = options || {};
+  var query = _.assign({}, options.query || {});
+  query.id = database.hash(id);
+  database.collections.jwtSecrets.update(
+    query, {
+      $set: {
+        'meta.updated': Date.now(),
+        'state': state
+      }
+    }, database.writeOptions, function(err, result) {
+      if(err) {
+        logger.error('Mongo error when trying to update proof state.', err);
+        return callback(new BedrockError(
+          'Failed to update proof state due to internal error.',
+          'InternalError',
+          {proof: id, httpStatusCode: 500, 'public': true},
+          err));
+      }
+      callback(null, result.result.n > 0);
+    });
+}
+
+/**
+ * Gets the cached proof state information, retrieving it from the database
+ * and recycling HMAC keys as needed.
+ *
+ * @param options the options to use
+ *   algorithm the JWT signing algorithm to use
+ *   clockToleranceInSecs the clockTolerance to use in during JWT verification
+ *   namespace the namespace to use
+ *   provision true for provisioning operation
+ *   ttlInSecs the JWT ttl
+ * @param callback(err, state) called once the operation completes.
+ */
+function _getState(options, callback) {
+  var done = callback;
+  async.auto({
+    get: function(callback) {
+      // use cached proof state if available, expiration checked later
```

Comment:
Blanket replace `proof state` with `state` or `key state` in comments as you see fit.